2D Moving wall shear-thnning lubrication

rbsecor opened the pull request:
First checkin of moving wall lubrication routines for shear-thinning liquids.  This version includes computing 3 intermediate variables "on the fly" whereas another option would be to incorporate those as Goma field variables.  The steady-state version of shear-thinning moving wall should work -- there still seem to be issues with the level-set imnplementation.

rbsecor commented:
Added some bounds on extraneous temperature values and hopefully made the shear-thinning shear-rate iteration more stable.  Hopefully this makes lub_energy run better.

rbsecor commented:
p.s. also, there are probably some conflicts with the Log_interpolation check-in -- will try to resolve those.

wortiz commented:
I'll look through it.

> On Jenkins, the step 8 results in a catastrophic failure -- will keep looking.

For Jenkins builds I have enabled FP_EXCEPT=ON which might be the reason.
```
Amesos_Mumps : ERROR
Amesos_Mumps : INFOG(1) = -10
Amesos_Mumps : INFOG(2) = 13355
Amesos_Mumps : On process 0, INFO(1) = -10
Amesos_Mumps : On process 0, INFO(2) = 13355
Amesos_Mumps : ERROR
Amesos_Mumps : INFOG(1) = -10
Amesos_Mumps : INFOG(2) = 13355
Amesos_Mumps : On process 0, INFO(1) = -10
Amesos_Mumps : On process 0, INFO(2) = 13355
[b5862d8e7c15:11816] *** Process received signal ***
[b5862d8e7c15:11816] Signal: Floating point exception (8)
[b5862d8e7c15:11816] Signal code: Floating point divide-by-zero (3)
[b5862d8e7c15:11816] Failing at address: 0x7bd7965c9568
[b5862d8e7c15:11816] [ 0] /lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7bd7929c8520]
[b5862d8e7c15:11816] [ 1] /lib/x86_64-linux-gnu/libm.so.6(feraiseexcept+0x38)[0x7bd7965c9568]
[b5862d8e7c15:11816] [ 2] /lib/x86_64-linux-gnu/libm.so.6(log10f64+0x40)[0x7bd7965cc390]
[b5862d8e7c15:11816] [ 3] ./goma.exe(solve_nonlinear_problem+0x5540)[0x573ef9660f40]
[b5862d8e7c15:11816] [ 4] ./goma.exe(hunt_problem+0x1d66)[0x573ef935fd86]
[b5862d8e7c15:11816] [ 5] ./goma.exe(main+0xd25)[0x573ef9359bc5]
[b5862d8e7c15:11816] [ 6] /lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7bd7929afd90]
[b5862d8e7c15:11816] [ 7] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7bd7929afe40]
[b5862d8e7c15:11816] [ 8] ./goma.exe(_start+0x25)[0x573ef935a6d5]
[b5862d8e7c15:11816] *** End of error message ***
--------------------------------------------------
```

wortiz commented:
The following patch fixes the floating point exception:

```
diff --git a/src/mm_sol_nonlinear.c b/src/mm_sol_nonlinear.c
index 4d01572b..87095fcb 100644
--- a/src/mm_sol_nonlinear.c
+++ b/src/mm_sol_nonlinear.c
@@ -1923,7 +1923,8 @@ int solve_nonlinear_problem(struct GomaLinearSolverData *ams,
     */

     if (inewton && (Resid_Norm_stack[2] > 0) && (Resid_Norm_stack[2] != 1) &&
-        (Resid_Norm_stack[1] > 0) && (Resid_Norm_stack[1] != 1) &&
+        (Resid_Norm_stack[1] > 0) && (Resid_Norm_stack[1] != 1) && (Soln_Norm_stack[2] > 0) &&
+        (Soln_Norm_stack[1] != 1) &&
         (inewton <= 1 || ((Resid_Norm_stack[0] > 0) && (Resid_Norm_stack[0] != 1)))) {
 #if 1
       if (inewton <= 1) {
@@ -1950,6 +1951,7 @@ int solve_nonlinear_problem(struct GomaLinearSolverData *ams,
     }
     if (nAC && inewton && (AC_Resid_Norm_stack[2] > 0) && (AC_Resid_Norm_stack[2] != 1) &&
         (AC_Resid_Norm_stack[1] > 0) && (AC_Resid_Norm_stack[0] != 1) &&
+        (AC_Soln_Norm_stack[2] > 0) && (AC_Soln_Norm_stack[1] != 1) &&
         (inewton <= 1 || ((AC_Resid_Norm_stack[0] > 0) && (AC_Resid_Norm_stack[0] != 1)))) {
 #if 1
       if (inewton <= 1 && DOUBLE_NONZERO(log10(AC_Resid_Norm_stack[1]))) {
```

wortiz commented:
`lub_CYmold` seems to be virtually the same, the absolute differences are around 1e-6, could probably be rebaselined unless we find some reason not too

`lub_energy` seems to have pretty big difference in convergence. It looks like a LUB_P with respect to the temperature issue.

Jacobian checker shows issues if you use a converged solution as initial guess.

It might be in lub_viscosity_integrate but not sure yet. I'll try to take a look again in the morning.

Previously:

```
	----------------------------------
	First Order Hunting:
	Step number:    2 of  200 (max)
	Attempting solution at: theta = 0.0521683 ;  step = 0.0521683
	MTID=  1 MPID= 4100 Parameter=  9.302326e-01 delta_s= -6.976744e-02
	MTID=  2 MPID= 4100 Parameter=  9.302326e-01 delta_s= -6.976744e-02
	BCID=  0 DFID=    1 Parameter= -3.188390e-02 delta_s= -2.920384e-03
	BCID=  0 DFID=    2 Parameter=  9.302326e-01 delta_s= -6.976744e-02

               R e s i d u a l         C o r r e c t i o n

   ToD   itn   L_oo    L_1     L_2     L_oo    L_1     L_2   lis  asm/slv (sec)  ROrd/Rt UOrd/Rt 
-------- --- ------- ------- ------- ------- ------- ------- --- --------------- ------- -------
18:22:35 [0] 2.2e+01 1.2e+02 4.4e+01 1.2e+00 1.6e+03 2.2e+01  1  5.5e-01/1.8e-01 0.0/0.0 0.0/0.0 
18:22:36 [1] 9.2e-01 5.6e+00 1.9e+00 2.0e-02 1.2e+01 2.1e-01  1  5.5e-01/1.7e-01 1.1/1.4 1.2/2.0 
18:22:36 [2] 9.5e-04 6.1e-03 2.0e-03 2.0e-05 5.5e-03 1.6e-04  1  5.6e-01/1.7e-01 2.2/3.8 1.6/3.7 
18:22:37 [3] 6.2e-10 4.8e-09 1.4e-09 4.8e-09 2.1e-05 2.7e-07  1  5.5e-01/1.8e-01 2.1/7.8 0.9/2.6 
-------- --- ------- ------- ------- ------- ------- ------- --- --------------- ------- -------
scaled solution norms   3.298e+03  1.186e+02  2.494e+02 
```

With this branch:

```
	----------------------------------
	First Order Hunting:
	Step number:    2 of  200 (max)
	Attempting solution at: theta = 0.0521683 ;  step = 0.0521683
	MTID=  1 MPID= 4100 Parameter=  9.302326e-01 delta_s= -6.976744e-02
	MTID=  2 MPID= 4100 Parameter=  9.302326e-01 delta_s= -6.976744e-02
	BCID=  0 DFID=    1 Parameter= -3.188390e-02 delta_s= -2.920384e-03
	BCID=  0 DFID=    2 Parameter=  9.302326e-01 delta_s= -6.976744e-02

               R e s i d u a l         C o r r e c t i o n

   ToD   itn   L_oo    L_1     L_2     L_oo    L_1     L_2   lis  asm/slv (sec)  ROrd/Rt UOrd/Rt 
-------- --- ------- ------- ------- ------- ------- ------- --- --------------- ------- -------
23:34:13 [0] 2.2e+01 1.4e+02 4.4e+01 1.2e+00 1.7e+03 2.2e+01  1  5.7e-01/1.9e-01 0.0/0.0 0.0/0.0 
23:34:13 [1] 1.4e+00 6.0e+01 5.5e+00 3.5e-01 1.9e+02 6.3e+00  1  5.6e-01/1.8e-01 1.1/0.9 1.1/0.5 
23:34:14 [2] 4.3e-02 2.3e+00 1.8e-01 6.2e-03 6.0e+00 1.0e-01  1  5.6e-01/1.8e-01 1.6/1.8 3.3/2.4 
23:34:15 [3] 1.2e-03 1.9e-02 2.2e-03 4.6e-05 2.0e-01 2.3e-03  1  5.6e-01/1.8e-01 1.3/2.1 0.9/1.6 
23:34:16 [4] 1.3e-05 2.2e-04 2.4e-05 5.7e-07 2.2e-03 2.5e-05  1  5.6e-01/1.8e-01 1.0/2.0 1.2/2.1 
23:34:16 [5] 1.9e-07 2.7e-06 3.6e-07 8.4e-09 2.8e-05 3.3e-07  1  5.6e-01/1.8e-01 0.9/1.8 1.0/1.9 
-------- --- ------- ------- ------- ------- ------- ------- --- --------------- ------- -------
scaled solution norms   3.157e+03  1.216e+02  2.436e+02 
```

rbsecor commented:
OK, It seems like we're getting closer.  Lots of small temperature sensitivity errors... I'll take another look at the dq/dT terms.

Eqdof=13348 LUBP_0 n=06675         Vardof=13351 SH_TEMP_0 n=06676      x=25         dx=4.6e-05    aj=-4.309e-10 nj=6.907e-06  aj_1=-4.3e-10   d_aj=6.1e-09    conf=23        
Eqdof=13349 SH_T_0 n=06675         Vardof=13351 SH_TEMP_0 n=06676      x=25         dx=4.6e-05    aj=0.00308    nj=0.003075   aj_1=0.0031     d_aj=0.0006     conf=4.4       
Eqdof=13350 LUBP_0 n=06676         Vardof=13351 SH_TEMP_0 n=06676      x=25         dx=4.6e-05    aj=8.483e-07  nj=2.388e-05  aj_1=8.5e-07    d_aj=1.2e-05    conf=45        
Eqdof=13352 LUBP_0 n=06677         Vardof=13351 SH_TEMP_0 n=06676      x=25         dx=4.6e-05    aj=3.641e-07  nj=3.606e-06  aj_1=3.6e-07    d_aj=2.1e-06    conf=12        
Eqdof=13353 SH_T_0 n=06677         Vardof=13351 SH_TEMP_0 n=06676      x=25         dx=4.6e-05    aj=-0.004534  nj=-0.004536  aj_1=-0.0045    d_aj=0.00012    conf=0.28      
Eqdof=13354 LUBP_0 n=06678         Vardof=13351 SH_TEMP_0 n=06676      x=25         dx=4.6e-05    aj=-4.043e-07 nj=-8.056e-06 aj_1=-4e-07     d_aj=2.9e-06    conf=18        
Eqdof=13355 SH_T_0 n=06678         Vardof=13351 SH_TEMP_0 n=06676      x=25         dx=4.6e-05    aj=-0.03644   nj=-0.03644   aj_1=-0.036     d_aj=0.00098    conf=5.1       
Diag        dof=13353 SH_T_0 n=06677 Var     dof=13353 SH_TEMP_0 n=06677 x=25         dx=4.636e-05  aj=0.08063    nj=0.08064    aj_1=0.08063    d_aj=2.515e-05  conf=1.619     
Eqdof=13262 LUBP_0 n=06632         Vardof=13353 SH_TEMP_0 n=06677      x=25         dx=4.6e-05    aj=4.905e-08  nj=3.983e-07  aj_1=4.9e-08    d_aj=7.5e-08    conf=0.71      
Eqdof=13281 SH_T_0 n=06641         Vardof=13353 SH_TEMP_0 n=06677      x=25         dx=4.6e-05    aj=-0.001283  nj=-0.001285  aj_1=-0.0013    d_aj=6.5e-06    conf=1.7       
Eqdof=13340 LUBP_0 n=06671         Vardof=13353 SH_TEMP_0 n=06677      x=25         dx=4.6e-05    aj=-1.19e-07  nj=-1.413e-06 aj_1=-1.2e-07   d_aj=2.3e-07    conf=5.4       
Eqdof=13342 LUBP_0 n=06672         Vardof=13353 SH_TEMP_0 n=06677      x=25         dx=4.6e-05    aj=5.936e-08  nj=4.484e-07  aj_1=5.9e-08    d_aj=7.7e-08    conf=0.65      
Eqdof=13349 SH_T_0 n=06675         Vardof=13353 SH_TEMP_0 n=06677      x=25         dx=4.6e-05    aj=-0.0247    nj=-0.0247    aj_1=-0.025     d_aj=4.6e-05    conf=0.23      
Eqdof=13355 SH_T_0 n=06678         Vardof=13353 SH_TEMP_0 n=06677      x=25         dx=4.6e-05    aj=-0.003452  nj=-0.003453  aj_1=-0.0035    d_aj=1.3e-06    conf=0.76      
Diag        dof=13355 SH_T_0 n=06678 Var     dof=13355 SH_TEMP_0 n=06678 x=25.37      dx=4.636e-05  aj=0.05098    nj=0.05101    aj_1=0.05098    d_aj=0.0001154  conf=34.11     
Eqdof=13262 LUBP_0 n=06632         Vardof=13355 SH_TEMP_0 n=06678      x=25         dx=4.6e-05    aj=-2.246e-07 nj=-1.804e-06 aj_1=-2.2e-07   d_aj=1.6e-06    conf=5         
Eqdof=13263 SH_T_0 n=06632         Vardof=13355 SH_TEMP_0 n=06678      x=25         dx=4.6e-05    aj=-0.01079   nj=-0.01079   aj_1=-0.011     d_aj=0.00031    conf=0.66

wortiz commented:
> OK, It seems like we're getting closer. Lots of small temperature sensitivity errors... I'll take another look at the dq/dT terms.

I haven't found the exact issue but I'm pretty sure that's the correct spot to be looking.

Also the first step is nexp=1 so that could hide some Jacobian issues for later steps if you don't change the continuation parameters.

wortiz commented:
The temperature field is quite a bit different in this version. It has a reasonable location for an increase in temperature with the viscous dissipation source, near the inflow and the hole seems like it should be a high shear area.

Not sure if something else is contributing to the change.

wortiz commented:
We might be able to get around the divergence issue by changing the continuation step size if the behavior looks more correct.

I think I (or someone) needs to add a Normalized Divergence Tolerance of like 1e6 where if the initial norm is increased by a factor of that tolerance we stop trying to iterate.

On this case the values become so large that we start to run into floating point exceptions when we could have failed the step.

wortiz commented:
.runguts

rbsecor commented:
Added some more nan guardrails and removed some code that seemed counterproductive.  Somewhere in this go-around, I did add shell_T sensitivities to the GRAD_LUB_PRESS bc -- which should arguably be there -- be it's possible for lub_energy this causes the initial residual to be off and more iterations on the first timestep.  (It seems the biggest residuals and soln updates are at the inflow plane.) One could also argue this shouldn't make any difference.  Anyway, that would be one argument for re-baselining lub_energy, assuming it still has some failed tests.

wortiz commented:
.runguts

wortiz commented:
For some reason the first step the number of Newton iterations changes in parallel I'm not sure if this is a mumps or goma issue. The differences in values computed don't really change just iterations taken. (for lub_energy)

I've reverted back to 1 processor.

wortiz commented:
By changes I mean one run is 5 iterations and another is 6 iterations but not sure why.
